ext/mini_racer_extension/mini_racer_extension.c

```diff
@@ -0,0 +1,1449 @@
+    return rb_utf8_str_new((char *)res.buf, res.len);
+}
+
+static VALUE context_pump_message_loop(VALUE self)
+{
+    Context *c;
+    Buf b;
+
+    TypedData_Get_Struct(self, Context, &context_type, c);
+    buf_init(&b);
+    buf_putc(&b, 'P');        // (P)ump, returns bool
+    return rendezvous(c, &b); // takes ownership of |b|
+}
+
+static VALUE context_idle_notification(VALUE self, VALUE arg)
+{
+    Context *c;
+    Ser s;
+
+    Check_Type(arg, T_FIXNUM);
+    TypedData_Get_Struct(self, Context, &context_type, c);
+    // request is (I)dle notification, idle_time_in_seconds
+    ser_init1(&s, 'I');
+    ser_num(&s, LONG2FIX(arg) / 1e3);
+    // response is |undefined|
+    return rendezvous(c, &s.b); // takes ownership of |s.b|
+}
+
+static VALUE context_low_memory_notification(VALUE self)
+{
+    Buf req, res;
+    Context *c;
+
+    TypedData_Get_Struct(self, Context, &context_type, c);
+    buf_init(&req);
+    buf_putc(&req, 'L');              // (L)ow memory notification, returns nothing
+    rendezvous_no_des(c, &req, &res); // takes ownership of |req|
+    return Qnil;
+}
+
+static int platform_set_flag1(VALUE k, VALUE v)
+{
+    char *p, buf[256];
+    int ok;
+
+    k = rb_funcall(k, rb_intern("to_s"), 0);
+    Check_Type(k, T_STRING);
+    if (!NIL_P(v)) {
+        v = rb_funcall(v, rb_intern("to_s"), 0);
+        Check_Type(v, T_STRING);
+    }
+    p = RSTRING_PTR(k);
+    if (!strncmp(p, "--", 2))
+        p += 2;
+    if (NIL_P(v)) {
+        snprintf(buf, sizeof(buf), "--%s", p);
+    } else {
+        snprintf(buf, sizeof(buf), "--%s=%s", p, RSTRING_PTR(v));
+    }
+    p = buf;
+    pthread_mutex_lock(&flags_mtx);
+    if (!flags.buf)
+        buf_init(&flags);
+    ok = (*flags.buf != 1);
+    if (ok)
+        buf_put(&flags, p, 1+strlen(p)); // include trailing \0
+    pthread_mutex_unlock(&flags_mtx);
+    return ok;
+}
+
+static VALUE platform_set_flags(int argc, VALUE *argv, VALUE klass)
+{
+    VALUE args, kwargs, k, v;
+    int i, n;
+
+    (void)&klass;
+    rb_scan_args(argc, argv, "*:", &args, &kwargs);
+    Check_Type(args, T_ARRAY);
+    for (i = 0, n = RARRAY_LENINT(args); i < n; i++) {
+        k = rb_ary_entry(args, i);
+        v = Qnil;
+        if (!platform_set_flag1(k, v))
+            goto fail;
+    }
+    if (NIL_P(kwargs))
+        return Qnil;
+    Check_Type(kwargs, T_HASH);
+    args = rb_ary_new();
+    rb_hash_foreach(kwargs, collect, args);
+    for (i = 0, n = RARRAY_LENINT(args); i < n; i += 2) {
+        k = rb_ary_entry(args, i+0);
+        v = rb_ary_entry(args, i+1);
+        if (!platform_set_flag1(k, v))
+            goto fail;
+    }
+    return Qnil;
+fail:
+    rb_raise(platform_init_error, "platform already initialized");
+}
+
+// called by v8_global_init
+void v8_get_flags(char **p, size_t *n)
+{
+    *p = NULL;
+    *n = 0;
+    pthread_mutex_lock(&flags_mtx);
+    if (!flags.len)
+        goto out;
+    *p = malloc(flags.len);
```

Comment:
are we leaking a p here? I am not seeing it get freed.

Reply:
It's freed by the caller. I've added a comment.